### PR TITLE
Bump REVEM-379 test coverage up up up.

### PR DESCRIPTION
**Pull request by davidjoy**
Also fixed a props "isRequired" issue that we missed earlier.

### CODE REVIEW

File: src/payment/cart/Offers.jsx [hunk 1]
```diff
@@ -60,7 +60,11 @@ function Offer({
 Offer.propTypes = {
   benefitType: PropTypes.oneOf(['Percentage', 'Absolute']).isRequired,
   benefitValue: PropTypes.number.isRequired,
-  provider: PropTypes.string.isRequired,
+  provider: PropTypes.string,
```
Comment (davidjoy) [inline src/payment/cart/Offers.jsx]: We missed this - now that provider isn't required, it was throwing an error here when we didn't provide it.

File: src/payment/data/sagas.js [hunk 2]
```diff
@@ -88,7 +88,7 @@ export function* handleDiscountCheck() {
         discount.jwt,
       );
       yield put(basketDataReceived(result));
-      yield call(handleMessages, result.messages, true);
+      yield call(handleMessages, result.messages, false);
```
Comment (davidjoy) [inline src/payment/data/sagas.js]: I tweaked this for efficiency - we were calling clearMessages() more often than we needed to.

File: src/payment/data/sagas.test.js [hunk 3]
```diff
@@ -103,6 +102,23 @@ describe('saga tests', () => {
   }
 
   describe('handleFetchBasket', () => {
+    it('should bail if the basket is processing', async () => {
```
Comment (davidjoy) [inline src/payment/data/sagas.test.js]: This test proves that we exit `catchBasketError` right away if the basket is processing.

File: src/payment/data/sagas.test.js [hunk 4]
```diff
@@ -148,12 +164,73 @@ describe('saga tests', () => {
         basketDataReceived(transformResults(response.data)),
         clearMessages(),
         basketDataReceived(transformResults(response.data)),
+        basketProcessing(false),
+        fetchBasket.fulfill(),
+      ]);
+      expect(caughtErrors).toEqual([]);
+      expect(mockApiClient.get).toHaveBeenCalledTimes(3);
+      expect(mockApiClient.get).toHaveBeenCalledWith(`${configuration.ECOMMERCE_BASE_URL}/bff/payment/v0/payment/`);
```
Comment (davidjoy) [inline src/payment/data/sagas.test.js]: Added this line to make a stronger expectation.

File: src/payment/data/sagas.test.js [hunk 5]
```diff
@@ -148,12 +164,73 @@ describe('saga tests', () => {
...
+      expect(mockApiClient.get).toHaveBeenCalledWith(
+        `${configuration.LMS_BASE_URL}/api/discounts/course/${courseKey}`,
+        {
+          xhrFields: { withCredentials: true },
+        },
+      );
+    });
+
+    it('should update basket data with jwt on second call', async () => {
```
Comment (davidjoy) [inline src/payment/data/sagas.test.js]: Added this test to prove that we use the discount_jwt parameter on the second getBasket call if it was provided by the discount check.

File: src/payment/data/sagas.test.js [hunk 6]
```diff
@@ -148,12 +164,73 @@ describe('saga tests', () => {
...
+        basketDataReceived(transformResults(response.data)),
         clearMessages(),
+        basketDataReceived(transformResults(response.data)),
         basketProcessing(false),
         fetchBasket.fulfill(),
       ]);
       expect(caughtErrors).toEqual([]);
+
       expect(mockApiClient.get).toHaveBeenCalledTimes(3);
+      expect(mockApiClient.get).toHaveBeenCalledWith(`${configuration.ECOMMERCE_BASE_URL}/bff/payment/v0/payment/`);
```
Comment (davidjoy) [inline src/payment/data/sagas.test.js]: Proving a bit more about the calls being made.

File: src/payment/data/sagas.test.js [hunk 7]
```diff
@@ -162,6 +239,39 @@ describe('saga tests', () => {
       );
     });
 
+    it('should update basket data without calling discount check API', async () => {
```
Comment (davidjoy) [inline src/payment/data/sagas.test.js]: Added this test to prove we didn't call the discount endpoint if we have no products.

File: src/payment/data/sagas.test.js [hunk 8]
```diff
@@ -311,8 +421,8 @@ describe('saga tests', () => {
       );
     });
 
-    it('should show a fallback error message', async () => {
-      response = { basket: Factory.build('basket', {}, { numProducts: 1 }) };
+    it('should update basket data and show an error message', async () => {
```
Comment (davidjoy) [inline src/payment/data/sagas.test.js]: Added this test to prove we call basketDataReceived with catchBasketError if we have a basket. and an error.

File: src/payment/cart/Offers.jsx (hunk 1 above)
Comment (MatthewPiatetsky) [reply src/payment/cart/Offers.jsx]: thanks!

**Pull request merged**